### PR TITLE
refactor(comm): change return of `send` to reflect the logic of delivery_group_size

### PR DESCRIPTION
**Pull request by Dylan-DPC**
change return of `send` to reflect the logic of delivery_group_size `delivery_group_size` or not

~~BREAKING CHANGE: return type of `Comm::send` is now `Result<SendStatus>`~~

### CODE REVIEW

**Review by dirvine — APPROVED** [review]

File: src/error.rs [hunk 1]
```diff
@@ -24,6 +25,14 @@ pub enum Error {
     FailedSignature,
     #[error("Cannot route.")]
     CannotRoute,
+    #[error("empty recipient list")]
```
Comment (maqi) [inline src/error.rs:28]: Empty

File: src/routing/comm.rs [hunk 2]
```diff
@@ -65,12 +64,16 @@ impl Comm {
         transport_config: qp2p::Config,
         event_tx: mpsc::Sender<ConnectionEvent>,
     ) -> Result<(Self, SocketAddr)> {
-        let quic_p2p = QuicP2p::with_config(Some(transport_config), &[], true)?;
+        let quic_p2p = QuicP2p::with_config(Some(transport_config), &[], true)
+            .map_err(|_| Error::InvalidConfig)?;
```
Comment (maqi) [inline src/routing/comm.rs:68]: but put an error log here to record the error from qp2p ?
same comment to the below

File: src/routing/comm.rs [hunk 3]
```diff
@@ -295,6 +291,15 @@ async fn handle_incoming_messages(
     }
 }
 
+/// Returns the status of the
+
```
Comment (maqi) [inline src/routing/comm.rs:295]: un-necessary line break?

File: src/routing/comm.rs [hunk 4]
```diff
@@ -335,9 +342,14 @@ mod tests {
...
-        comm.send(&[peer0.addr, peer1.addr], 1, message.clone())
-            .await
-            .0?;
+        let status = comm
+            .send(&[peer0.addr, peer1.addr], 1, message.clone())
+            .await?;
+
+        assert_matches!(
+            status,
+            SendStatus::AllRecipients | SendStatus::MinDeliveryGroupSizeReached(_)
```
Comment (maqi) [inline src/routing/comm.rs:351]: expected result shall be just AllRecipients? why `MinDeliveryGroupSizeReached` is also expected?

File: src/routing/comm.rs [hunk 5]
```diff
@@ -364,9 +376,16 @@ mod tests {
...
 
         let message = Bytes::from_static(b"hello world");
-        let (result, failed_recipients) = comm.send(&[invalid_addr], 1, message.clone()).await;
-        assert!(result.is_err());
-        assert_eq!(failed_recipients, [invalid_addr]);
+        let status = comm.send(&[invalid_addr], 1, message.clone()).await?;
+
+        assert_matches!(
+            &status,
+            &SendStatus::MinDeliveryGroupSizeFailed(_)
```
Comment (maqi) [inline src/routing/comm.rs:383]: can you do `SendStatus::MinDeliveryGroupSizeFailed([invalid_addr])` , and avoid the below if clause?

File: src/routing/comm.rs [hunk 6]
```diff
@@ -410,13 +429,19 @@ mod tests {
...
 
-        assert!(result.is_err());
-        assert_eq!(failed_recipients, [invalid_addr]);
-        assert_eq!(peer.rx.recv().await, Some(message));
+        assert_matches!(
+            status,
+            SendStatus::MinDeliveryGroupSizeFailed(_)
+        );
+
+        if let SendStatus::MinDeliveryGroupSizeFailed(failed_recipients) = status.clone() {
```
Comment (maqi) [inline src/routing/comm.rs:441]: is this clone really needed?

File: src/routing/core.rs [hunk 7]
```diff
@@ -1108,7 +1108,7 @@ impl Core {
             };
             if recipients.is_empty() {
                 trace!("Cannot route user message, recipient list empty: {:?}", msg);
-                return Err(Error::CannotRoute);
+                return Err(Error::EmptyRecipientList);
```
Comment (maqi) [inline src/routing/core.rs:1111]: I do think the previous `CannotRoute` error back is more accurate?

File: src/routing/dispatcher.rs [hunk 8]
```diff
@@ -190,15 +191,24 @@ impl Dispatcher {
...
-        let cmds = match message {
-            MessageType::Ping | MessageType::NodeMessage(_) => self
-                .comm
-                .send(recipients, delivery_group_size, msg_bytes)
-                .await
-                .1
-                .into_iter()
-                .map(Command::HandlePeerLost)
-                .collect(),
+        let cmds: Vec<Command> = match message {
```
Comment (maqi) [inline src/routing/dispatcher.rs:194]: really need `: Vec<Command>` ?

File: src/routing/core.rs (hunk 7 above)
Comment (Dylan-DPC) [reply src/routing/core.rs:1111]: I felt `EmpyRecipientList` gives a more clearer image of the error compared to `CannotRoute`

**Review by maqi — APPROVED** [review]

**Pull request merged**